### ISSUE TITLE
QA: params/vars and loops in dvc.yaml (2.0) round 2

### ISSUE THREAD

**Issue by jorgeorpinel** [iterative/dvc#5165]
> Continues #4854

### Bugs?

- ~~`dvc repro` executes stage groups alphabetically e,g, if my `foreach` list is `foo, bar, baz` (in that order) still the stages gets reproed (and saved to dvc.locl) in this order: `stg@bar`, `stg@baz`, `stg@foo`.~~
  Generalized and extracted to #5181

### UI/UX

- Foreach loops: since `do:` always contains all the regular stage fields, what about avoiding `do` and go straight to `cmd`, etc.? Like this:
  ```yaml
  stages:
    something:
      foreach: ...
      cmd: python script.py --thresh ${item...}
      deps: ...
  ```
  > This way the only syntax change is to add the `foreach` field, which makes the whole entry into a stage group.
- [x] If a directory path in `vars`, repro prints a generic `ERROR: unexpected error - [Errno 21] Is a directory: '{path}'.` which while informative enough, seems like it could be better handled. (Fix on #5171)

### Feature design/ Side-effects

- There seems to be no way to indicate you don't want any params file to be loaded (even if it exists for other purposes). Earlier on you could `use: none`.
- Similarly, there's no way to ONLY use an alternative params file e.g. `params.json` and avoid the default (`params.yaml`). I feel like the default behavior should be to skip params.yaml if any params files are specified.
- ~~`vars` cannot overwrite any values in params files, but params files accept repeated objects IF they can be merged without conflict.~~ E.g.: params.json `{"grp": {"a": 1}}` params2.json `{"grp": {"b": 2}}` can both be included even when `grp` is overloaded, ~~but you can't define `grp` at all in `vars`.~~

**Comment by jorgeorpinel** [comment]
@skshetry this is still WIP but you can start taking a look at this point if you'd like. Thanks

**Comment by skshetry** [comment]
> dvc repro executes stage groups alphabetically e,g, if my foreach list is foo, bar, baz (in that order) still the stages gets reproed (and saved to dvc.locl) in this order: stg@bar, stg@baz, stg@foo

Not a bug. The ordering is based on the DAG. Other than that, it might be difficult to promise to order (and, not that it matters). Lock file has its own way of ordering things wherever it matters, it does not have to work the way you expect.

> Foreach loops: since do: always contains all the regular stage fields, what about avoiding do and go straight to cmd, etc.?

The reasoning for this is that we have two levels of vars: local and global. Three if you count the foreach's `key` and `value`, but they come from global vars most of the time.

`do` helps them separate. Not in a clean way that I'd like, as `wdir` depends on global variables rather than local variables as mentioned in https://github.com/iterative/dvc.org/pull/2052#discussion_r549063671.

> ... the whole entry into a stage group.

I don't like this term at all. They are not a *group* in any way. If you are confusing it with `dvc repro <foreach_name>`, then that's purely a convenience function.

> There seems to be no way to indicate you don't want any params file to be loaded (even if it exists for other purposes). (Earlier on you could use: none.)

I brought that in the discussion with @dmpetrov. The main reasoning was that most people will use `params.yaml` by default, and if they do use other files, that either means they use separate files like `params.json` and don't use `params.yaml` or they do use `params.yaml` for common configs but use other files for specialized params.

> vars cannot overwrite any values in params files, but params files accept repeated objects

`vars` has the same behaviour whether you use files or declare it in `vars` itself. It never allows you to overwrite, you can only recursively load them. So, the following should work in `vars`:
```yaml
vars:
  grp:
    b: 3
```

This is useful in the top-level `vars`, but rarely useful in the `stage`-level `vars` (where, `key: value` is a better choice). 

We have decided that if someone wants to have an "overwrite" method, we'd introduce a new keyword as follows in the future:
```yaml
overwrite_vars: true
vars:
  grp: 3
```

**Comment by jorgeorpinel** [comment]
> The ordering is based on the DAG

So if there are no DAG connections in a stage group, why use any other order than what is specified in dvc.yaml? We even assign 0...n indices to some lists given to `foreach`. That seems much more intuitive IMO but @efiop @shcheklein can you please weight in?

> it does not have to work the way you expect.

DVC should not work in a predictable/intuitive way? I probably didn't understand what you meant with that @skshetry.

> they come from global vars most of the time

I think we are assuming there. OR do we know that users actually prefer global vs. local vars?

> do helps them separate

I don't have a string opinion on this since it's already implemented that way but I think it looks cleaner without `do`, and the change vs. regular stages is minimal: you just add the `foreach:` field to the spec. It could even be called something else like `set:`. Again, something for others to weight in, hopefully.

> They are not a group in any way

![image](https://user-images.githubusercontent.com/1477535/103315471-46dd9100-49eb-11eb-81a9-2f891f8b53c3.png)
> https://www.merriam-webster.com/dictionary/group

I see you fixed the "repro prints a generic ERROR" bug, great!

> > no way to indicate you don't want any params file
>
> most people will use params.yaml... if you don't want any parameterization, don't use it at all... If you only want to declare variables in vars section only, then name it a bit differently

No strong opinion either but those are workarounds. What if I want parameterization with `vars` with the same names as what is in params.yaml? Since it doesn't let me overwrite them, `use: none` was useful for that. I'm not saying that exact syntax should be brought back, maybe just `- none`. Up to you guys, I'm just leaving all the QA results here.

> no way to ONLY use an alternative params file e.g. params.json... (Maybe) the default behavior should be to skip params.yaml if any params files are specified?

Is there a reasoning for that? Thanks

> vars has the same behaviour whether you use files or declare it in vars itself

You're right! I confirmed this does work consistently 👍 (except that `grp` in your example should be a `-` list item).

**Comment by shcheklein** [comment]
> why use any other order than what is specified in dvc.yaml? 

any additional guarantee reduces flexibility (let's say we decide to run them in parallel at some point and people already expect certain order). I would wait for someone to ask this, for a specific use case to start considering this at all.

> but I think it looks cleaner without do:

I think @skshetry mentioned that `do` is needed. Not sure what do you suggest in terms of avoiding it?

> They are not a group in any way

so, according to that screenshot I tend to agree with @skshetry - since they do not represent a _unit_ in any way

**Comment by jorgeorpinel** [comment]
> any additional guarantee reduces flexibility (let's say we decide to run them in parallel

@shcheklein Let's use a simple example to make sure we're talking about the same thing:

```yaml
stages:
  notagroup:
    foreach:
     - 1
     - 2
     - 3
     cmd: echo ${item}
```

(BTW notice I also used my suggested syntax without `do:` above (previously mentioned in https://github.com/iterative/dvc.org/pull/2052#discussion_r549483306). I don't see why it's needed but again, up to you guys. In fact I've already documented it with `do`.)

So you're saying maybe we'll introduce some new field signaling to run them in parallel? OK that's fine, in that case you would be specifying an order too. My point is that if there's no way to build a DAG nor an order gets specified, the intuitive behavior would be to use the explicit order in which dvc.yaml is written.

Why does that reduce flexibility? It's just a predictable outcome. At least I'd like to know what that order is because it's not random. In fact I've noticed that changing the command or other parts of regular stages (not linked by DAGs) can result in a different execution order — so this seems to be an existing/general situation BTW: we can def. extract it from here if needed.

**Comment by shcheklein** [comment]
@jorgeorpinel 

> I don't see why it's needed

hmm ... what about the explanation @skshetry provided, with global/local vars? 

> So you're saying maybe we'll introduce some new field signaling to run them in parallel?

no. We might run them in parallel by default, nothing prevents us from doing this at the moment.

> My point is that if there's no way to build a DAG nor an order gets specified ...

I'm not sure I understand this to be honest. Could you clarify please?

> Why does that reduce flexibility?

you kinda set specific thing (execution order in this case). This is by definition reduces flexibility - we have to maintain that specification. E.g. it makes harder to run stuff in parallel by default.

> In fact I've noticed that changing the command or other parts of regular stages

Yes. We don't guarantee any order.

**Comment by jorgeorpinel** [comment]
> They are not a group in any way
> they do not represent a unit in any way

@shcheklein I can see that you and @skshetry feel very strongly about this so to clarify: I'm not married to the term group but I really can't think of anything more precise for what this feature does or how it looks from the user perspective, at least:

```yaml
stages:
  notaunit:  # one stage entry (an element in a whole) = unit
    foreach: # assemblage of (the items below) = group
     - 1     # object
     - 2     # object
     do:
       cmd: echo ${item}
```

Perhaps you guys can explain what you mean when you use definitive statements like "its' not that at all", why you think it's such a wrong thing, or provide other alternatives? TBH it's kind of difficult to QA without help from you to understand the intentions behind the design of the feature. Part of the point of someone outside of the eng. team doing so is to get a fresh perspective...

**Comment by jorgeorpinel** [comment]
p.s. to me "group" is simply "more than one thing". Could also be called "a set" I guess. But if the idea is to keep the "loop" and "foreach" terminology I think that wouldn't be user-friendly at all: "Stage loops"? "Foreach in stages"?

**Comment by shcheklein** [comment]
@jorgeorpinel it's a language construct to generate stages to my mind. Not sure how else I would describe it. As I mentioned I would analyze how users describe/ask about it in the first place- and would use that language (it should be correct though, e.g. we don't like group, because group is a very specific term in math/languages, and in our case it's not a group, these stages are not different from any other stages. Also, you can take a look at some other tools that use loops in yaml.

Writing this, I see where this confusion comes from. There are two languages now - `dvc.yaml` and `dvc.lock`. On the `dvc.yaml` level it's indeed looks like a single "entity", but on the `dvc.lock` level - it's a bunch of stages, not connected to each other in any special way.

I'm btw would be totally fine to use "group" in some informal way. The misleading would be to make it formal (use in title, and everywhere else). At least we should about some better term here.

**Comment by jorgeorpinel** [comment]
> it's not a group, these stages are not different from any other stages

OK I think see what your concern is now.

> I would analyze how users describe/ask about it in the first place

Definitely agree on this. I'm just not sure where to find that info...

> on the dvc.lock level - it's a bunch of stages, not connected

They're indirectly connected by the stage name prefixes e.g. `mystg@1`, `mystg@2`, etc. and always printed next to each other.

p.p.s. I'm not sure why this terminology discussion is happening here (it's not something I reported related to QA). Let's move this part to https://github.com/iterative/dvc.org/pull/2052#pullrequestreview-559754434 please!

**Comment by jorgeorpinel** [comment]
@shcheklein answers to the other comments (which I just noticed):

> > I don't see why `do:` is needed
>
> hmm ... what about the explanation @skshetry provided, with global/local vars?

TBH I didn't understand that explanation. Here's an example with all 3 types of vars andwithout `do`:

```yaml
vars:
- foo: foo

stages:
  mystages:
    vars:
    - bar: bar
    foreach:
    - 1
    - 2
    cmd: echo ${foo} ${bar} ${item}
```

I will create a separate issue on the order of stage execution and summarize/clarify there, since it seems to be more general than the scope of this QA ticket.

**Comment by jorgeorpinel** [comment]
OK, created #5181 for the order thing.

The only matter still in discussion here is about the `do` syntax, which as I've said from the beginning is not major from my PoV. Please consider my latest comment and example on that and also, feel free to close this ticket if the decision is to keep it.

I'm also building a 3rd (hopefully last) QA ticket around 2.0 pipelines: #5180

**Comment by jorgeorpinel** [comment]
I lied, there's 2 other questions still pending:
- no way to indicate you don't want any params file (previously `use: none`)
- no way to ONLY use an alternative params file (skip/overwrite params.yaml)

**Comment by jorgeorpinel** [comment]
My proposal would be to address both (not necessarily a priority) with something like

```yaml
vars:
- none
- foor: bar # This can now "override" values in params.yaml
- params.json # This will be the only params file loaded.
```

**Comment by jorgeorpinel** [comment]
Or (for the latter one) perhaps if you specify params files then by default DVC shouldn't load params.yaml (so no `none` needed in that case)

**Comment by jorgeorpinel** [comment]
Closing in favor of #5312
